apps/clock_info/lib.js

```diff
@@ -284,38 +284,49 @@ exports.addInteractive = function(menu, options) {
     E.stopEventPropagation&&E.stopEventPropagation();
   }
   Bangle.on("swipe",swipeHandler);
+  const blur = () => {
+    options.focus=false;
+    delete Bangle.CLKINFO_FOCUS;
+    options.redraw();
+    const itm = menu[options.menuA].items[options.menuB];
+    if (itm.blur) itm.blur(options);
+  };
+  const focus = (redraw) => {
+    let shown = false;
+    if (!options.focus) {
+      options.focus=true;
+      Bangle.CLKINFO_FOCUS=true;
+      shown = true;
+    }
+    if (redraw) options.redraw();
+    if (shown) {
+      const itm = menu[options.menuA].items[options.menuB];
+      if (itm.focus) itm.focus(options);
+    }
+  };
   let touchHandler, lockHandler;
   if (options.x!==undefined && options.y!==undefined && options.w && options.h) {
     touchHandler = function(_,e) {
       if (e.x<options.x || e.y<options.y ||
           e.x>(options.x+options.w) || e.y>(options.y+options.h)) {
-        if (options.focus) {
-          options.focus=false;
-          delete Bangle.CLKINFO_FOCUS;
-          options.redraw();
-        }
+        if (options.focus)
+          blur();
         return; // outside area
       }
       if (!options.focus) {
-        options.focus=true; // if not focussed, set focus
-        Bangle.CLKINFO_FOCUS=true;
-        options.redraw();
+        focus(true);
       } else if (menu[options.menuA].items[options.menuB].run) {
         Bangle.buzz(100, 0.7);
-        menu[options.menuA].items[options.menuB].run(); // allow tap on an item to run it (eg home assistant)
+        menu[options.menuA].items[options.menuB].run(options); // allow tap on an item to run it (eg home assistant)
       } else {
-        options.focus=true;
-        Bangle.CLKINFO_FOCUS=true;
+        focus();
```

Comment:
I'm not sure if this is needed here - `options.focus` is already `true`, was it just to set `CLKINFO_FOCUS` ?

Reply:
Good spot! It looks like this was added by @glemco to support quicklaunch/widhid - but it seems `options.focus=true` was in there already at that point, so this it does seem redundant.

We could take it out but we'd have to change the other apps, and I guess they are kind of independent of the clock_info module, so they'd have to do something like:

```
var clockinfo_focus = false;
try {
  clockinfo_focus = require("clock_info").clockInfos.some(i=>i.focus);
} catch { /* no module */ }
```

To catch the case that the module wasn't installed, which is not ideal I guess.

Reply:
Cool, and `options.focus = true` goes all the way back to the [introduction of focus](https://github.com/bobrippling/BangleApps/commit/11f63df3b1337c576701e8ee823d5745e1013749#diff-00d918fab679ddd859bdcd79e6c3e86474e1c74f8a25cc9c64d36784517f5146R220), so I reckon it's safe to remove both `focus` & `CLKINFO_FOCUS` being set here (or now, the call to `focus()`).

I don't understand the removal / other apps depending on this - `options.focus` is true at that point, so taking out the assignment won't have behaviour any apps can observe, right?

Reply:
What @gfwilliams wanted to point out is that the global `CLKINFO_FOCUS` is independent on the clkinfo module (not the cleanest thing I have to admit). In this way apps like quicklaunch and widhid can check it without needing to know if the module is installed or not.
The main idea behind needing it in quicklaunch is that the app uses swipes to do actions, now if the clkinfo is focused, the swipe would both go to the next clkinfo and open the action from quicklaunch, which is not exactly what a user wants.
If `options.focus` carries the same information we can indeed use that, but there seems to be no better way than using that try/catch .
We could have something like a global variable (or handler/lib, you pick what is cleanest) where apps register themselves as the ones needing exclusive swipe control, other apps would check that and in this way we are not directly dependent on clkinfo and the functionality can be extended to other apps needing it.
Could be something as simple as:
```javascript
Bangle.EXCLUSIVE_SWIPE == null; // no app registered
Bangle.EXCLUSIVE_SWIPE == 'clkinfo'; // the clkinfo module is currently registered
function registerExclusiveSwipe (app) {
  Bangle.EXCLUSIVE_SWIPE = app;
}
function IsSwipeExclusive () {
  return Bangle.EXCLUSIVE_SWIPE == null;
}
```

Reply:
Ah yes, I think there was misunderstanding on my part in that case, in this part of code, both `options.focus` and `Bangle.CLKINFO_FOCUS` will already be true, so there's no need to set them both to true again here (see 0943d202), but we need to preserve `CLKINFO_FOCUS`, which is done.

Slightly related (and I'd move this to another PR), but we could have [`E.stopPropagation`](https://github.com/espruino/Espruino/commit/35c8cb9be11ce76ea1829a4a8b40f2ab051bb60b) in the clkinfo handlers now, which would avoid other apps needing to check `CLKINFO_FOCUS` at all, what do you both think?

Reply:
That would be much neater actually, quite a nice function, I wasn't aware of it

Reply:
> , but we could have E.stopPropagation in the clkinfo handlers now, which would avoid other apps needing to check CLKINFO_FOCUS at all, what do you both think?

This works if the clock info handler is before other handlers in the related handler list. Otherwise other handlers will be called before we get to the clock info. IIrc. @gfwilliams

Reply:
> Slightly related (and I'd move this to another PR), but we could have [E.stopPropagation](https://github.com/espruino/Espruino/commit/35c8cb9be11ce76ea1829a4a8b40f2ab051bb60b) in the clkinfo handlers now, which would avoid other apps needing to check CLKINFO_FOCUS at all, what do you both think?

Yes, that sounds great. I think keeping `.focus` itself would be useful, but it'd be good to get rid of `CLKINFO_FOCUS`

As @thyttan says, we'd need to ensure we were first in the listener list, but I think we could handle that nicely using `prependListener` and then only adding `swipeHandler` when we're focussed (rather than always, like we do currently)